File: crates/hana-fs/src/policy.rs
//! Scan limits.
//!
//! This mirrors `hana_engine.indexing.policy.ScanPolicy` field for field. The
//! shared invariants are written down in `docs/SCAN_CONTRACT.md`; changing one
//! side without the other is a bug.

use serde::{Deserialize, Serialize};

/// Directories skipped unless the user removes them from the project settings.
pub const DEFAULT_IGNORED_DIRECTORIES: &[&str] = &[
    ".git",
    "node_modules",
    "target",
    "dist",
    "build",
    ".next",
    ".venv",
    "venv",
    "__pycache__",
    ".idea",
    ".vscode",
    "coverage",
];

/// Filename patterns skipped by default. Only a leading `*` wildcard is
/// supported, which covers every pattern the engine actually ships.
pub const DEFAULT_IGNORED_FILES: &[&str] = &[
    "*.pyc", "*.pyo", "*.class", "*.o", "*.so", "*.dll", "*.exe", "*.lock",
    ".DS_Store", "Thumbs.db",
];

/// 5 MiB.
pub const DEFAULT_MAX_FILE_SIZE_BYTES: u64 = 5 * 1024 * 1024;

/// Depth in path components below the project root.
pub const DEFAULT_MAX_DEPTH: usize = 24;

/// Bytes read when deciding whether a file is binary.
pub const BINARY_SNIFF_BYTES: usize = 8192;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanPolicy {
    pub ignored_directories: Vec<String>,
    pub ignored_files: Vec<String>,
    pub max_file_size_bytes: u64,
    pub max_depth: usize,
    /// Off by default: following a symlink is how a scan escapes the folder the
    /// user chose. Even when enabled, the resolved target must still be inside
    /// the project root.
    pub follow_symlinks: bool,
    pub hash_binary_files: bool,
    #[serde(default)]
    pub extra_ignored_directories: Vec<String>,
}

impl Default for ScanPolicy {
    fn default() -> Self {
        Self {
            ignored_directories: DEFAULT_IGNORED_DIRECTORIES
                .iter()
                .map(|s| (*s).to_string())
                .collect(),
            ignored_files: DEFAULT_IGNORED_FILES
                .iter()
                .map(|s| (*s).to_string())
                .collect(),
            max_file_size_bytes: DEFAULT_MAX_FILE_SIZE_BYTES,
            max_depth: DEFAULT_MAX_DEPTH,
            follow_symlinks: false,
            hash_binary_files: true,
            extra_ignored_directories: Vec::new(),
        }
    }
}

impl ScanPolicy {
    pub fn ignores_directory(&self, name: &str) -> bool {
        self.ignored_directories.iter().any(|item| item == name)
            || self
                .extra_ignored_directories
                .iter()
                .any(|item| item == name)
    }

    pub fn ignores_file(&self, name: &str) -> bool {
        self.ignored_files
            .iter()
            .any(|pattern| matches_pattern(pattern, name))
    }
}

/// Match a filename against a pattern supporting a single leading `*`.
///
/// Kept intentionally small: the default list only needs suffix matching, and a
/// full glob engine would be one more dependency to audit.
fn matches_pattern(pattern: &str, name: &str) -> bool {
    match pattern.strip_prefix('*') {
        Some(suffix) => name.ends_with(suffix),
        None => pattern == name,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_ignore_list_matches_the_specification() {
        let policy = ScanPolicy::default();
        for name in [
            ".git",
            "node_modules",
            "target",
            "dist",
            "build",
            ".next",
            ".venv",
            "venv",
            "__pycache__",
            ".idea",
            ".vscode",
            "coverage",
        ] {
            assert!(policy.ignores_directory(name), "{name} should be ignored");
        }
        assert!(!policy.ignores_directory("src"));
    }

    #[test]
    fn suffix_patterns_match() {
        let policy = ScanPolicy::default();
        assert!(policy.ignores_file("module.pyc"));
        assert!(policy.ignores_file(".DS_Store"));
        assert!(!policy.ignores_file("guardar.sql"));
    }

    #[test]
    fn extra_ignores_are_honoured() {
        let policy = ScanPolicy {
            extra_ignored_directories: vec!["legacy".to_string()],
            ..Default::default()
        };
        assert!(policy.ignores_directory("legacy"));
    }

    #[test]
    fn round_trips_through_json_like_the_engine_sends_it() {
        let policy = ScanPolicy::default();
        let json = serde_json::to_string(&policy).unwrap();
        let parsed: ScanPolicy = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.max_depth, DEFAULT_MAX_DEPTH);
        assert!(!parsed.follow_symlinks);
    }
}

File: engine/hana_engine/indexing/policy.py
"""Scan policy — the single definition of what HANA is allowed to read.

Both scanners (the Rust one in ``crates/hana-fs`` and the Python one next
door) implement this policy. It is written down once, here and in
``docs/SCAN_CONTRACT.md``, so the two cannot drift silently.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from typing import Any

#: Directories skipped unless the user removes them from the project settings.
DEFAULT_IGNORED_DIRECTORIES: tuple[str, ...] = (
    ".git",
    "node_modules",
    "target",
    "dist",
    "build",
    ".next",
    ".venv",
    "venv",
    "__pycache__",
    ".idea",
    ".vscode",
    "coverage",
)

#: Filename globs skipped by default.
DEFAULT_IGNORED_FILES: tuple[str, ...] = (
    "*.pyc",
    "*.pyo",
    "*.class",
    "*.o",
    "*.so",
    "*.dll",
    "*.exe",
    "*.lock",
    ".DS_Store",
    "Thumbs.db",
)

#: 5 MiB. Larger source files are inventoried but not hashed or analysed.
DEFAULT_MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024

#: Depth is counted in path components below the project root.
DEFAULT_MAX_DEPTH = 24

#: Bytes read when deciding whether a file is binary.
BINARY_SNIFF_BYTES = 8192


@dataclass(frozen=True)
class ScanPolicy:
    """Limits enforced on every directory walk.

    ``follow_symlinks`` defaults to False. Following a symlink is how a scan
    escapes the folder the user selected, so turning it on is an explicit,
    per-project decision — and even then the resolved target must still fall
    inside the project root.
    """

    ignored_directories: tuple[str, ...] = DEFAULT_IGNORED_DIRECTORIES
    ignored_files: tuple[str, ...] = DEFAULT_IGNORED_FILES
    max_file_size_bytes: int = DEFAULT_MAX_FILE_SIZE_BYTES
    max_depth: int = DEFAULT_MAX_DEPTH
    follow_symlinks: bool = False
    hash_binary_files: bool = True
    extra_ignored_directories: tuple[str, ...] = field(default=())

    def __post_init__(self) -> None:
        # A policy now arrives from the settings screen, so these are reachable
        # by typing rather than only by a programming mistake. A zero size limit
        # or a zero depth would scan nothing at all and look like an empty
        # project — a wrong answer is worse than a rejected one.
        if self.max_file_size_bytes <= 0:
            raise ValueError("max_file_size_bytes debe ser mayor que cero")
        if self.max_depth <= 0:
            raise ValueError("max_depth debe ser mayor que cero")

    @property
    def all_ignored_directories(self) -> frozenset[str]:
        return frozenset(self.ignored_directories) | frozenset(
            self.extra_ignored_directories
        )

    def with_overrides(self, **changes: Any) -> ScanPolicy:
        return replace(self, **changes)

    @classmethod
    def from_dict(cls, data: Mapping[str, Any] | None) -> ScanPolicy:
        """Build a policy from persisted project settings, ignoring junk keys."""
        if not data:
            return cls()
        return cls(
            ignored_directories=tuple(
                data.get("ignored_directories", DEFAULT_IGNORED_DIRECTORIES)
            ),
            ignored_files=tuple(data.get("ignored_files", DEFAULT_IGNORED_FILES)),
            max_file_size_bytes=int(
                data.get("max_file_size_bytes", DEFAULT_MAX_FILE_SIZE_BYTES)
            ),
            max_depth=int(data.get("max_depth", DEFAULT_MAX_DEPTH)),
            follow_symlinks=bool(data.get("follow_symlinks", False)),
            hash_binary_files=bool(data.get("hash_binary_files", True)),
            extra_ignored_directories=tuple(data.get("extra_ignored_directories", ())),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "ignored_directories": list(self.ignored_directories),
            "ignored_files": list(self.ignored_files),
            "max_file_size_bytes": self.max_file_size_bytes,
            "max_depth": self.max_depth,
            "follow_symlinks": self.follow_symlinks,
            "hash_binary_files": self.hash_binary_files,
            "extra_ignored_directories": list(self.extra_ignored_directories),
        }
